### PR TITLE
UHM-4534 Add Lab ID data entry/view to Lab Result Entry page

### PR DESCRIPTION
**Pull request by brandones**
https://pihemr.atlassian.net/browse/UHM-4534

Screenshot from 2020-05-21 16-45-01
Screenshot from 2020-05-21 17-36-47

labid

~~Known limitation: if you go to a lab with status Reported or Collected, enter a Lab ID straightaway, then hit "Edit", then immediately hit "Submit", the Lab ID will not be submitted. Not sure whether this is a problem or not.~~

The accession number is shown when in view mode. A text box is shown when in edit mode.

### CODE REVIEW

File: app/js/components/LabResultEntry.jsx [hunk 1]
```diff
@@ -10,11 +10,12 @@ import React, { PureComponent } from 'react';
 import { connect } from 'react-redux';
 import PropTypes from 'prop-types';
 import R from 'ramda';
-import { formValueSelector, change, untouch } from 'redux-form';
+import { formValueSelector, change } from 'redux-form';
```
Comment (brandones) [inline app/js/components/LabResultEntry.jsx:13]: unrelated cleanup

File: app/js/components/LabResultEntry.jsx [hunk 2]
```diff
@@ -69,7 +73,7 @@ export class LabResultEntry extends PureComponent {
       labResultsDidNotPerformReasonQuestion,
       labResultsTestLocationQuestion,
     } = this.props;
-    if (typeof state !== 'undefined') {
+    if (state) {
```
Comment (brandones) [inline app/js/components/LabResultEntry.jsx:76]: unrelated cleanup

File: app/js/components/LabResultEntry.jsx [hunk 3]
```diff
@@ -540,60 +575,90 @@ export class LabResultEntry extends PureComponent {
...
+          {location.state && (
             <div>
               <h2 className="lab-entry-page-title">
                 <FormattedMessage
                   id="app.labResultEntry.title"
-                  defaultMessage="Test Results -" />
-                {` ${getConceptShortName(location.state.concept,locale)}`}
+                  defaultMessage="Test Results -"
+                />
+                {` ${getConceptShortName(location.state.concept, locale)}`}
```
Comment (brandones) [inline app/js/components/LabResultEntry.jsx:585]: I meant to only auto-format my work, but maybe more of the file got hit...

File: app/js/components/LabResultEntry.jsx [hunk 4]
```diff
@@ -540,60 +575,90 @@ export class LabResultEntry extends PureComponent {
...
-                        <span className="test-details">{moment(location.state.dateActivated).format('MMM DD h:mm A')}</span>
+                          defaultMessage="Order Date:"
+                        />{" "}
+                        <span className="test-details">
+                          {moment(location.state.dateActivated).format(
+                            "MMM DD h:mm A"
+                          )}
+                        </span>
+                      </span>
+                    </div>
```
Comment (brandones) [inline app/js/components/LabResultEntry.jsx:645]: Nothing but formatting, and a couple bootstrap column size changes

File: app/js/components/LabResultEntry.jsx [hunk 5]
```diff
@@ -605,14 +670,18 @@ export class LabResultEntry extends PureComponent {
...
               <br />
               <br />
-              <button className="cancel" type="button" onClick={() => window.location.assign(returnUrl)}>Return</button>
+              <button
+                className="cancel"
+                type="button"
+                onClick={() => window.location.assign(returnUrl)}
+              >
+                Return
+              </button>
```
Comment (brandones) [inline app/js/components/LabResultEntry.jsx:684]: formatting

File: app/js/components/LabResultEntry.jsx [hunk 6]
```diff
@@ -91,6 +98,32 @@ export class LabResultEntry extends PureComponent {
...
       this.reloadCachedValues();
     }
+    if (!this.state.isFormSubmitListenerAttached) {
+      // This component's existing form handling is both highly abstracted and
+      // highly inflexible. We needed to add a type of input to it that doesn't
+      // fit its structure. Rather than attempt to hack *through* all of that, we
+      // are just hacking *onto* it.
+      //
+      // We attach an onclick event listener to the submit button. It makes an
+      // additional API call.
```
Comment (mogoodrich) [inline app/js/components/LabResultEntry.jsx:108]: cool, thanks for the explanation!

Comment (mogoodrich) [reply app/js/components/LabResultEntry.jsx:108]: And I like the approach.

**Comment by mogoodrich** [comment]
Thanks @brandones !
